File: code/349.intersection-of-two-arrays.go
/*
 * @lc app=leetcode id=349 lang=golang
 *
 * [349] Intersection of Two Arrays
 */

// @lc code=start
func intersection(nums1 []int, nums2 []int) []int {
	set := make(map[int]bool)
	for _, v := range nums1 {
		set[v] = true
	}

	var res []int

	for _, v := range nums2 {
		if set[v] {
			res = append(res, v)
			delete(set, v)
		}
	}

	return res
}

// @lc code=end

File: code/349.intersection-of-two-arrays.java
/*
 * @lc app=leetcode id=349 lang=java
 *
 * [349] Intersection of Two Arrays
 */

// @lc code=start

import java.util.HashSet;
import java.util.Set;

class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        if (nums1.length == 0 || nums2.length == 0) {
            return new int[0];
        }
        Set<Integer> set = new HashSet<>();
        for (int num: nums1) {
            set.add(num);
        }

        Set<Integer> res = new HashSet<>();
        for (int num: nums2) {
            if (set.contains(num)) {
                res.add(num);
            }
        }
        return res.stream().mapToInt(Integer::intValue).toArray();
    }
}
// @lc code=end
